src/peer.rs

```diff
@@ -8,90 +8,100 @@
+use futures::StreamExt;
 use tokio::task::JoinHandle;
+use libp2p::{
+    core::{muxing::StreamMuxerBox, transport::Boxed},
+    dns::TokioDnsConfig,
+    gossipsub,
+    identify, identity::{self}, noise,
+    swarm::{SwarmBuilder, SwarmEvent},
+    tcp, yamux, PeerId, Transport,
+};
+use libp2p_mplex::MplexConfig;
+use pchain_types::cryptography::PublicAddress;
+use pchain_types::blockchain::TransactionV1;
+use std::net::Ipv4Addr;
+use std::time::Duration;
 
-use crate::config::Config;
-use crate::messages::{DroppedTxnMessage, Message, Topic};
-
-/// The builder struct for constructing a [Peer].
-pub struct PeerBuilder {
-    /// Configurations of the peer
-    pub config: Config,
+use crate::{
+    behaviour::{Behaviour, NetworkEvent},
+    conversions,
+    messages::{DroppedTxnMessage,Message, Topic},
+    config::Config,
+};
 
-    /// Message handler to process received messages from the network
-    pub handlers: Vec<Box<dyn Fn(PublicAddress, Message) + Send>>,
-}
 
-impl PeerBuilder {
-    pub fn new(config: Config) -> PeerBuilder {
-        PeerBuilder {
-            config,
-            handlers: vec![],
-        }
-    }
-
-    pub fn on_receive_msg(
-        mut self,
-        handler: impl Fn(PublicAddress, Message) + Send + 'static,
-    ) -> PeerBuilder {
-        self.handlers.push(Box::new(handler));
-        self
-    }
-
-    /// Constructs a [Peer] from the given configuration and handlers, and start the thread for the p2p network.
-    pub async fn build(self) -> Result<Peer, EngineError> {
-        crate::engine::start(self).await
-    }
-}
 
 pub struct Peer {
-    /// Network handle for the [tokio::task] which is the main thread for the
-    /// p2p network (see [crate::engine]).
-    pub(crate) engine: JoinHandle<()>,
+    /// Network handle for the [tokio::task] which is the main thread for the p2p network 
+    pub(crate) handle: JoinHandle<()>,
 
-    /// mpsc sender for delivering messages to the p2p network.
-    pub(crate) to_engine: tokio::sync::mpsc::Sender<EngineCommand>,
+    /// mpsc sender for delivering [PeerCommand] to the p2p network.
+    pub(crate) sender: tokio::sync::mpsc::Sender<PeerCommand>,
 }
 
 impl Peer {
+/// Constructs a [Peer] from the given configuration and handlers and start the thread for the p2p network \
+/// 1. Load network configuration to set up transport for the P2P network. 
+/// 2. Establishes connection to the network by adding bootnodes and subscribing to message [Topic]. 
+/// 3. Spawns an asynchronous [tokio] task and enters the main event loop, returning a mpsc Sender used for sending 
+/// [PeerCommand] to the internal thread.
+///
+    pub async fn start(config: Config, handlers: Vec<Box<dyn Fn(PublicAddress, Message) + Send>>) -> Result<Peer, PeerStartError> {
+        let swarm = set_up_transport(&config).await?;
+        let swarm = establish_network_connections(swarm, &config)?;
+        let (handle, sender) = main_loop(swarm, &config, handlers)?;
```

Comment:
"main_loop" doesn't seem to be a good naming...?

Reply:
i agree, maybe "spawn_event_loop" , "start_event_detection" ? any suggestions?